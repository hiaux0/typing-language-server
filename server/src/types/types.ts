import { Position } from "vscode-languageserver-textdocument";

/**
 * file://path/to/file.ext
 */
export type Uri = string;

/**
 * Type from the client: vscode.TextEditor.Selection
 */
export interface ClientEditorSelection {
  /**
   * The position at which the selection starts.
   * This position might be before or after [active](#Selection.active).
   */
  anchor: Position;
  /**
   * The position of the cursor.
   * This position might be before or after [anchor](#Selection.anchor).
   */
  active: Position;
  start: Position;
  end: Position;

  /**
   * A selection is reversed if [active](#Selection.active).isBefore([anchor](#Selection.anchor)).
   */
  // isReversed: boolean;
}

export interface GetEditorSelectionResponse {
  documentText: string;
  documentUri: string;
  documentPath: string;
  selections: ClientEditorSelection[];
}

export interface TypoAnalytics {
  text: string;
  mispelled: number;
}

export interface TypingAnalytics {
  occurrence: number;
  typos: TypoAnalytics[];
}

// export type AnalyticsMap = Map<string, Analytics>;
export type TypingAnalyticsMap = Record<string, TypingAnalytics>;

export type TypingLessons =
  | "words"
  | "alphabet" // abcdefghijklmnopqrstuvwxyz
  | "alphabet-chunks" // chunks of alphabet
  | "ac" // alphabet-chunks
  | "bigrams" // todo
  | "vim"; // todo
export interface WordsFilterConfigurationInput {
  lesson?: TypingLessons;
  amount: number; // amount of words
  length: number; // length of the word
  repeat?: number;
  autoEnter?: boolean; // Auto new line when finished. Note: Either this or clearLineOnFinish
  autoNewWords?: boolean;
  clearOnError?: boolean;
  clearLineOnFinish?: boolean; // Note: Either this or autoEnter
  // letters: string | string[], // "rs" or ["st", "rs"]
  ignore?: string | string[]; // "rs" or ["st", "rs"]
  anyOrder?: string | string[]; // "rs, at" or ["rs", "at"]
  oneOf?: string | string[]; // "rs" or ["r", "s"]
  sequence?: string | string[]; // "rs, at" or ["rs", "at"]
}

export interface WordsFilterConfigurationOutput
  extends Required<WordsFilterConfigurationInput> {
  ignore: string[]; // "rs" or ["st", "rs"]
  oneOf: string[]; // ["r", "s"]
  sequence: string[]; // ["r", "s"]
}
