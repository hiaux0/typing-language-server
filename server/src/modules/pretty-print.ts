import { AnalyticsMap } from "../types/types";
import { getWordsFromAnalytics } from "./analytics";
import { findLongest } from "./string";

export function prettyPrintTypoTable(map: AnalyticsMap, word: string | undefined): string[] {
    if (!word) return []
    const data = map[word]
    const words = getWordsFromAnalytics(map, word);
    const longest = findLongest(words);
    const cellLength = longest?.length + 4 + 2 + 1; // + 4: ` - <num>` ;; + 2: space before and after the `|` in markdown borders ;; + 1: space after the longest word
    const firstLine = `| ${word} - ${data.occurrence}`.padEnd(cellLength, ' ');
    const others = []
    for (const typo of data.typos) {
        others.push(`| ${typo.text} - ${typo.mispelled}`.padEnd(cellLength, ' '));
    }
    const separator = `|${'-'.repeat(cellLength - 1)}`;
    const result = [firstLine, separator, ...others]
    return result
}

// const map: AnalyticsMap = {
//     "abcde": {
//         "occurrence": 1,
//         "typos": [
//             {
//                 "text": "ax",
//                 "mispelled": 2
//             },
//             {
//                 "text": "abcdex",
//                 "mispelled": 3
//             }
//         ]
//     }
// }
// const word = "abcde"
// const result = prettyPrintTypoTable(map, word)
// console.log("[pretty-print.ts,55] result: ", result);


export function prettyPrintTypoTableAll(map: AnalyticsMap): string[] {
    // 1. get each row prettified
    const prettyfiedColumns = Object.keys(map).map((word) => {
        return prettyPrintTypoTable(map, word);
    });
    // 2. join the rows from top to bottom
    const result = []
    let iterations = 0
    let rowIndex = 0
    let elementExists = true
    while (elementExists && iterations < 7) {
        elementExists = prettyfiedColumns.some(col => !!col[rowIndex])
        const joinedRow = []
        // 2.1 Go from top to bottom
        for (const col of prettyfiedColumns) {
            if (!col[rowIndex]) {
                // 2.1.1 add empty rows to keep table structure
                joinedRow.push('|' + ' '.repeat(col[0].length - 1));
                continue
            }
            joinedRow.push(col[rowIndex])
        }
        result.push(joinedRow.join('') + '|')
        iterations++
        rowIndex++
    }
    result.pop() // one extra empty row is added in while>for>if>continue
    return result;
}

// const map: AnalyticsMap = {
//     "abcde": {
//         "occurrence": 1,
//         "typos": [
//             {
//                 "text": "ax",
//                 "mispelled": 2
//             },
//             {
//                 "text": "abcdex",
//                 "mispelled": 3
//             }
//         ]
//     },
//     "123": {
//         "occurrence": 1,
//         "typos": [
//             {
//                 "text": "1x",
//                 "mispelled": 2
//             },
//             // {
//             //     "text": "01234x",
//             //     "mispelled": 3
//             // }
//         ]
//     }
// };
// const result = prettyPrintTypoTableAll(map)
// console.log(result[0]);
// console.log(result[1]);
// console.log(result[2]);
// console.log(result[3]);
// console.log(result[4]);

