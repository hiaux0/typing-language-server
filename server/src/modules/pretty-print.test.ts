import { AnalyticsMap } from "../types/types";
import { prettyPrintTypoTable, prettyPrintTypoTableAll } from "./pretty-print";

describe('pretty-print.ts', () => {
    describe('prettyPrintTypoTable', () => {
        test('1 entry', () => {
            const map: AnalyticsMap = {
                "abcde": {
                    "occurrence": 1,
                    "typos": [
                        {
                            "text": "ax",
                            "mispelled": 2
                        },
                        {
                            "text": "abcdex",
                            "mispelled": 3
                        }
                    ]
                }
            }
            const word = "abcde"
            const result = prettyPrintTypoTable(map, word)
            expect(result).toMatchSnapshot()
        })
        test('2 entries', () => {
            const map: AnalyticsMap = {
                "abcde": {
                    "occurrence": 1,
                    "typos": [
                        {
                            "text": "ax",
                            "mispelled": 2
                        },
                        {
                            "text": "abcdex",
                            "mispelled": 3
                        }
                    ]
                },
                "123": {
                    "occurrence": 1,
                    "typos": [
                        {
                            "text": "1x",
                            "mispelled": 2
                        },
                        // {
                        //     "text": "01234x",
                        //     "mispelled": 3
                        // }
                    ]
                }
            }
            const result = prettyPrintTypoTableAll(map)
            expect(result).toMatchSnapshot()
        })
    });
});
