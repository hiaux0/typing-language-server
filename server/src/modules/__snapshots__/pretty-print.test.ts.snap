// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pretty-print.ts prettyPrintTypoTable 1 entry 1`] = `
[
  "| abcde - 1  ",
  "|------------",
  "| ax - 2     ",
  "| abcdex - 3 ",
]
`;

exports[`pretty-print.ts prettyPrintTypoTable 2 entries 1`] = `
[
  "| 123 - 1 | abcde - 1  |",
  "|---------|------------|",
  "| 1x - 2  | ax - 2     |",
  "|         | abcdex - 3 |",
]
`;
